src/Common/DateLUTImpl.h

```diff
@@ -381,19 +381,19 @@ class DateLUTImpl
     {
         const LUTIndex i = toLUTIndex(v);
         if constexpr (std::is_unsigned_v<DateOrTime> || std::is_same_v<DateOrTime, DayNum>)
-            return lut_saturated[i - lut[i].day_of_month + lut[i].days_in_month].date;
+            return lut_saturated[i + (lut[i].days_in_month - lut[i].day_of_month)].date;
         else
-            return lut[i - lut[i].day_of_month + lut[i].days_in_month].date;
+            return lut[i + (lut[i].days_in_month - lut[i].day_of_month)].date;
```

Comment:
> Furthermore the "expensive" LUT index arithmetic is performed twice, for example in toLastDayOfMonth:
first time i - lut[i].day_of_month and the second time when we add lut[i].days_in_month.

Maybe I miss something, but would adressing the unnecessary computation not mean that `lut[i]` is pulled out? E.g.

```cpp
const LUTIndex i = toLUTIndex(v);
const Value & lut_day = lut[i];
if constexpr (std::is_unsigned_v<DateOrTime> || std::is_same_v<DateOrTime, DayNum>)
  return lut_saturated[i + lut_day.days_in_month - lut_day.day_of_month].date;
...
```

Reply:
By "expensive" I mean for example `operator +`:

```
    template <typename T>
    friend inline LUTIndex operator+(const LUTIndex & index, const T v)
    {
        return normalizeLUTIndex(index.toUnderType() + UInt32(v));
    }

    static inline LUTIndex normalizeLUTIndex(UInt32 index)
    {
        if (index >= DATE_LUT_SIZE)
            return LUTIndex(DATE_LUT_SIZE - 1);
        return LUTIndex{index};
    }

    static inline LUTIndex normalizeLUTIndex(Int64 index)
    {
        if (unlikely(index < 0))
            return LUTIndex(0);
        if (index >= DATE_LUT_SIZE)
            return LUTIndex(DATE_LUT_SIZE - 1);
        return LUTIndex{index};
    }

```
So it's better to not use less LUT operators whenever it's possible, for example, it's better first add or subtract numbers and then add result to LUT index.

The same already done in other functions of master branch. For example:
```
    template <typename DateOrTime>
    inline Time toFirstDayOfWeek(DateOrTime v) const
    {
        const LUTIndex i = toLUTIndex(v);
        if constexpr (std::is_unsigned_v<DateOrTime> || std::is_same_v<DateOrTime, DayNum>)
            return lut_saturated[i - (lut[i].day_of_week - 1)].date;
        else
            return lut[i - (lut[i].day_of_week - 1)].date;
    }
```
So contributors who implemented toLastDayOfMonth and other functions in this PR simply forgot to use the parentheses and the correct numbers order.

Reply:
Got it, thanks. I didn't notice `operator+` on `LUTIndex`es...